### ISSUE TITLE
New check: unpinned repo clone

### ISSUE THREAD

**Issue by laurentsimon** [ossf/scorecard#608]
**Is your feature request related to a problem? Please describe.**
detect `git`, `hg` and other source management tools used to clone repo without pinning by hash.
This check is problematic because there are many legitimate reason to do this, especially for one's own repo. Example, for integration tests, etc.
Maybe we could detect repo clones that are different from the repo itself. Note that dependabot probably does not support updating the version of a cloned repo.

**Describe the solution you'd like**
need discussion
**Describe alternatives you've considered**
need discussion

**Additional context**

**Comment by laurentsimon** [comment]
possible patterns to check for:

1. Git/hg repo clon
2. Curl github.com/bla/bla/releases/download/bla
3. Curl tarball

**Comment by ig33** [comment]
Support all

**Comment by chrismcgehee** [comment]
Here's my opinion: we should not do unpinned dependency checks for dependency types not supported by dependabot. If a developer modifies their code to clone a git repo at a certain commit hash, chances are that hash will not be updated for years, if ever. Yes, they will be protected from someone pushing malicious code to that git repo in the future, but they will also miss out on vulnerability and other bug fixes. Without dependabot, I would say the developer would be less secure by pinning a hash.

**Comment by laurentsimon** [comment]
it would be good to flag it nevertheless, because it's a potential risk. Any thoughts on warning without recommending pinning by hash?

**Comment by chrismcgehee** [comment]
Yeah, I think if we just noted this as a warning, but didn't recommend pinning, that would be a good approach.

**Comment by laurentsimon** [comment]
do you know if dependabot/renovatebot support commands within shell script/makefiles `go install bla@something`? I'm also curious about GCB's cloud.yaml files https://github.com/ossf/scorecard/issues/1503

cc @rarkins

**Comment by rarkins** [comment]
Right now we don't support arbitrary scripts unless you're self hosting. For the app it's too high risk

**Comment by laurentsimon** [comment]
@rarkins do you mean it's too risky to parse arbitrary shell scripts?

Let me give you a concrete example of what I meant.
if I have a workflow using `uses: some/action@somehash`, renovate-bot is able to send PR to update the hash when a new action of released.
If I use `run: go install somepackage@hash`, is renovaote-bot able to understand that I depend on `somepackage`, even though it's part in a `go.sum` file?

More generally, is removate-bot capable of understanding cloud.yaml depedencies given via args, or commands such as `git reset --hard SOME_HASH`. (This one clearly looks harder)

**Comment by rarkins** [comment]
Sorry, I misunderstood.

Renovate attempts to understand as many patterns as it can out of the box, but it doesn't know those ones. For ones like that which it doesn't understand, users can add regex patterns to identify them for Renovate. The feature is know as the "regex manager": https://docs.renovatebot.com/modules/manager/regex/

**Comment by laurentsimon** [comment]
Interesting, thanks for the info.
